Refactor out QKV in projection

erichan1 opened the pull request:
Summary: Refactor to reduce amount of copied code for decoder by finding common chunks for encoder and decoder. QKV in projection is a reasonable unit to copy out.

Test Plan:
buck run mode/opt -c fbcode.platform=platform010 -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 //pytext/fb/tools:benchmark_transformers -- transformer --batch-size 64 --avg-sequence-length 235 --max-sequence-length 256 --iters 100  --module native

Benchmark and numerical tests work fine.

Reviewed By: mikekgfb

Differential Revision: D36138504

vadimkantorov commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -282,6 +282,50 @@ std::tuple<Tensor, Tensor, Tensor> transform_bias_rescale_qkv_cpu(
+        q_k_v_weight_s.size() == 3,
+        "expected chunk to produce 3 tensors but it produced ",
+        q_k_v_weight_s.size());
+    // TODO: can we stay nested instead of using cat?
```
I think nested now triggers a cat internally anyway :( @cpuhrsch

erichan1 replied:
Good to know. A later TODO.

jbschlosser commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -282,6 +282,50 @@ std::tuple<Tensor, Tensor, Tensor> transform_bias_rescale_qkv_cpu(
   return std::make_tuple(q_k_v_s[0], q_k_v_s[1], q_k_v_s[2]);
 }
 
+Tensor qkv_matmul(
```
naming nit: something like `qkv_projection` is a bit more specific

jbschlosser reviewed (commented):
LGTM! One minor nit, nothing blocking

erichan1 committed:
Refactor out QKV in projection (#79437)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/79437

Refactor to reduce amount of copied code for decoder by finding common chunks for encoder and decoder. QKV in projection is a reasonable unit to copy out.

Test Plan:
buck run mode/opt -c fbcode.platform=platform010 -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 //pytext/fb/tools:benchmark_transformers -- transformer --batch-size 64 --avg-sequence-length 235 --max-sequence-length 256 --iters 100  --module native

Benchmark and numerical tests work fine.

Reviewed By: mikekgfb

Differential Revision: D36138504

fbshipit-source-id: 1832f0d94282cde008977f3ce1225bde3139c957

erichan1 commented:
@pytorchbot merge -g

erichan1 commented:
@pytorchbot merge -g
